### ISSUE TITLE
A bug in the current implementation - clusterClient not closed properly.

### ISSUE THREAD

**Issue by FuyaoLi2017** [wangyang0918/flink-native-k8s-operator#11]
Hello Yang,
I noticed a problem in the current implementation.
In ` https://github.com/wangyang0918/flink-native-k8s-operator/blob/master/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkApplicationController.java#L270`

and
`https://github.com/wangyang0918/flink-native-k8s-operator/blob/master/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkApplicationController.java#L241`

You are actually creating a new `clusterClient` without closing it. There will be a lot of stucked connections that will eventually make the container reach error code 137 and restart. This process might take 10~20 hours.

I understand `clusterClient` has implemented the `AutoClosable` interface. However, we can't directly use try-with-resources to recycle the resource since the trigger savepoint action has a nested `CompletableFuture`. Wrap it inside the try-with-resources will break this logic and gives you errors like this.

```
2021-09-21 00:26:27,960 WARN  org.apache.flink.shaded.netty4.io.netty.channel.AbstractChannel [] - Force-closing a channel whose registration task was not accepted by an event loop: [id: 0xadcdcf10]
java.util.concurrent.RejectedExecutionException: event executor terminated
	at org.apache.flink.shaded.netty4.io.netty.util.concurrent.SingleThreadEventExecutor.reject(SingleThreadEventExecutor.java:926) ~[flink-native-k8s-operator-1.0-SNAPSHOT.jar:?]
	at org.apache.flink.shaded.netty4.io.netty.util.concurrent.SingleThreadEventExecutor.offerTask(SingleThreadEventExecutor.java:353) ~[flink-native-k8s-operator-1.0-SNAPSHOT.jar:?]
	at org.apache.flink.shaded.netty4.io.netty.util.concurrent.SingleThreadEventExecutor.addTask(SingleThreadEventExecutor.java:346) ~[flink-native-k8s-operator-1.0-SNAPSHOT.jar:?]
	at org.apache.flink.shaded.netty4.io.netty.util.concurrent.SingleThreadEventExecutor.execute(SingleThreadEventExecutor.java:828) ~[flink-native-k8s-operator-1.0-SNAPSHOT.jar:?]
	at org.apache.flink.shaded.netty4.io.netty.util.concurrent.SingleThreadEventExecutor.execute(SingleThreadEventExecutor.java:818) ~[flink-native-k8s-operator-1.0-SNAPSHOT.jar:?]
	at org.apache.flink.shaded.netty4.io.netty.channel.AbstractChannel$AbstractUnsafe.register(AbstractChannel.java:471) ~[flink-native-k8s-operator-1.0-SNAPSHOT.jar:?]
	at org.apache.flink.shaded.netty4.io.netty.channel.SingleThreadEventLoop.register(SingleThreadEventLoop.java:87) ~[flink-native-k8s-operator-1.0-SNAPSHOT.jar:?]
	at org.apache.flink.shaded.netty4.io.netty.channel.SingleThreadEventLoop.register(SingleThreadEventLoop.java:81) ~[flink-native-k8s-operator-1.0-SNAPSHOT.jar:?]
	at org.apache.flink.shaded.netty4.io.netty.channel.MultithreadEventLoopGroup.register(MultithreadEventLoopGroup.java:86) ~[flink-native-k8s-operator-1.0-SNAPSHOT.jar:?]
	at org.apache.flink.shaded.netty4.io.netty.bootstrap.AbstractBootstrap.initAndRegister(AbstractBootstrap.java:323) ~[flink-native-k8s-operator-1.0-SNAPSHOT.jar:?]
	at org.apache.flink.shaded.netty4.io.netty.bootstrap.Bootstrap.doResolveAndConnect(Bootstrap.java:155) ~[flink-native-k8s-operator-1.0-SNAPSHOT.jar:?]
	at org.apache.flink.shaded.netty4.io.netty.bootstrap.Bootstrap.connect(Bootstrap.java:139) ~[flink-native-k8s-operator-1.0-SNAPSHOT.jar:?]
	at org.apache.flink.shaded.netty4.io.netty.bootstrap.Bootstrap.connect(Bootstrap.java:123) ~[flink-native-k8s-operator-1.0-SNAPSHOT.jar:?]
	at org.apache.flink.runtime.rest.RestClient.submitRequest(RestClient.java:421) ~[flink-native-k8s-operator-1.0-SNAPSHOT.jar:?]
	at org.apache.flink.runtime.rest.RestClient.sendRequest(RestClient.java:344) ~[flink-native-k8s-operator-1.0-SNAPSHOT.jar:?]
	at org.apache.flink.runtime.rest.RestClient.sendRequest(RestClient.java:258) ~[flink-native-k8s-operator-1.0-SNAPSHOT.jar:?]
	at org.apache.flink.client.program.rest.RestClusterClient.lambda$null$23(RestClusterClient.java:777) ~[flink-native-k8s-operator-1.0-SNAPSHOT.jar:?]
	at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1072) [?:?]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) [?:?]
	at java.util.concurrent.CompletableFuture.postFire(CompletableFuture.java:610) [?:?]
	at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:649) [?:?]
	at java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:478) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.lang.Thread.run(Thread.java:829) [?:?]
2021-09-21 00:26:27,968 ERROR org.apache.flink.shaded.netty4.io.netty.util.concurrent.DefaultPromise.rejectedExecution [] - Failed to submit a listener notification task. Event loop shut down?
java.util.concurrent.RejectedExecutionException: event executor terminated
	at org.apache.flink.shaded.netty4.io.netty.util.concurrent.SingleThreadEventExecutor.reject(SingleThreadEventExecutor.java:926) ~[flink-native-k8s-operator-1.0-SNAPSHOT.jar:?]
	at org.apache.flink.shaded.netty4.io.netty.util.concurrent.SingleThreadEventExecutor.offerTask(SingleThreadEventExecutor.java:353) ~[flink-native-k8s-operator-1.0-SNAPSHOT.jar:?]
	at org.apache.flink.shaded.netty4.io.netty.util.concurrent.SingleThreadEventExecutor.addTask(SingleThreadEventExecutor.java:346) ~[flink-native-k8s-operator-1.0-SNAPSHOT.jar:?]
	at org.apache.flink.shaded.netty4.io.netty.util.concurrent.SingleThreadEventExecutor.execute(SingleThreadEventExecutor.java:828) ~[flink-native-k8s-operator-1.0-SNAPSHOT.jar:?]
	at org.apache.flink.shaded.netty4.io.netty.util.concurrent.SingleThreadEventExecutor.execute(SingleThreadEventExecutor.java:818) ~[flink-native-k8s-operator-1.0-SNAPSHOT.jar:?]
	at org.apache.flink.shaded.netty4.io.netty.util.concurrent.DefaultPromise.safeExecute(DefaultPromise.java:841) ~[flink-native-k8s-operator-1.0-SNAPSHOT.jar:?]
	at org.apache.flink.shaded.netty4.io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:498) ~[flink-native-k8s-operator-1.0-SNAPSHOT.jar:?]
	at org.apache.flink.shaded.netty4.io.netty.util.concurrent.DefaultPromise.addListener(DefaultPromise.java:183) ~[flink-native-k8s-operator-1.0-SNAPSHOT.jar:?]
	at org.apache.flink.shaded.netty4.io.netty.channel.DefaultChannelPromise.addListener(DefaultChannelPromise.java:95) ~[flink-native-k8s-operator-1.0-SNAPSHOT.jar:?]
	at org.apache.flink.shaded.netty4.io.netty.channel.DefaultChannelPromise.addListener(DefaultChannelPromise.java:30) ~[flink-native-k8s-operator-1.0-SNAPSHOT.jar:?]
	at org.apache.flink.runtime.rest.RestClient.submitRequest(RestClient.java:425) ~[flink-native-k8s-operator-1.0-SNAPSHOT.jar:?]
	at org.apache.flink.runtime.rest.RestClient.sendRequest(RestClient.java:344) ~[flink-native-k8s-operator-1.0-SNAPSHOT.jar:?]
	at org.apache.flink.runtime.rest.RestClient.sendRequest(RestClient.java:258) ~[flink-native-k8s-operator-1.0-SNAPSHOT.jar:?]
	at org.apache.flink.client.program.rest.RestClusterClient.lambda$null$23(RestClusterClient.java:777) ~[flink-native-k8s-operator-1.0-SNAPSHOT.jar:?]
	at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1072) [?:?]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) [?:?]
	at java.util.concurrent.CompletableFuture.postFire(CompletableFuture.java:610) [?:?]
	at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:649) [?:?]
	at java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:478) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.lang.Thread.run(Thread.java:829) [?:?]

```

I tried to add a Thread.sleep() in the try body part and I can get rid of this error. However, the savepoint completion time is not very deterministic due to various reasons. Doing a Thread.sleep() is not the optimal solution and if there is an exception. The close() statement in the try body won't catch it.

I am not sure how we can solve this elegantly? I would like to hear your input on this. Thanks!

**Comment by FuyaoLi2017** [comment]
I can patch a MR for this later.

**Comment by wangyang0918** [comment]
Already addressed in #12
